Cache routes.url_helpers built modules

jhawthorn opened the pull request:
Previously, every time this was called it would return a new module.

Every time a new controller is built, it had one of these module included onto it (see [`AbstractController::Railties::RoutesHelper`](https://github.com/rails/rails/blob/master/actionpack/lib/abstract_controller/railties/routes_helpers.rb)). Because each time this was a new module we would end up with a different copy being included on each controller. This would also include a different copy on a controller than on its superclass (ex. `ApplicationController`). Furthermore, each generated module also extended the url helper modules.

```
    +--------------+     +-------------+
    |              |     |             |
    | path_helpers |     | url_helpers | (named routes added here)
    |              |     |             |
    +------+----+--+     +----+--+-----+
           ^    ^             ^  ^
           |    |             |  |
           |  +---------------+  |
           |  | |                |
           |  | +-----------+    |
           |  |             |    |
    +------+--+---+      +--+----+-----+
    |             |      |             |
    | (singleton) +------+ url_helpers | (duplicated for each controller)
    |             |      |             |
    +-------------+      +-----+-------+
                               ^
                               |
                               |
                     +---------+-------+
                     |                 |
                     | UsersController |
                     |                 |
                     +-----------------+
```

The result of this is that when named routes were added to the two top-level modules, defining those methods became extremely slow because Ruby has to invalidate the class method caches on all of the generated modules as well as each controller. Furthermore because there were multiple paths up the inheritance chain (ex. one through UsersController's generated module and one through ApplicationController's generated module) it would end up invaliding the classes multiple times (ideally Ruby would be smarter about this, but it's much easier to solve in Rails).

Caching this module means that all controllers should include the same module and **defining these named routes should be 3-4x faster**. One caveat: this is only measurable with eager loading on, since the controllers being loaded is what causes the slow method definitions.

This method can generate two different modules, based on whether supports_path is truthy, so those are cached separately.

## Benchmark

``` ruby
require "action_controller/railtie"
require "rails"

class MyApp < Rails::Application
  config.eager_load = true
  config.secret_key_base = SecureRandom.hex
  initialize!
end

class ApplicationController < ActionController::Base
end

N = (ENV["N"] || 500).to_i
M = (ENV["M"] || 2000).to_i
N.times do |n|
  Object.const_set("Foo#{n}Controller", Class.new(ApplicationController))
end
puts "For #{N} controllers:"

MyApp.routes.draw do
  before = RubyVM.stat(:class_serial)
  get "/foo123", to: "home#index"
  after = RubyVM.stat(:class_serial)
  puts "  Class serial changed by #{after - before} per route"

  ms = Benchmark.ms {
    M.times do |n|
      get "/foo#{n}", to: "home#index"
    end
  }
  puts "  Defining #{M} routes took #{ms}ms"
end
```

**Before:**
```
For 500 controllers:
  Class serial changed by 13543 per route
  Defining 2000 routes took 1069.1329431720078ms
```

**After**
```
For 500 controllers:
  Class serial changed by 2028 per route
  Defining 2000 routes took 309.9443819373846ms
```

cc @tenderlove who helped ❤️
cc @casperisfine @Edouard-chin @rafaelfranca who were looking into `define_url_helper`'s performance in #37139. I'd love to know if this helps you too 🙇

jhawthorn committed:
Cache routes.url_helpers built modules

Previously, every time this was called it would return a new module.

Every time a new controller is built, it had one of these module
included onto it (see AbstractController::Railties::RoutesHelper).
Because each time this was a new module we would end up with a different
copy being included on each controller. This would also include a
different copy on a controller than on its superclass (ex.
ApplicationController). Furthermore, each generated module also extended
the url helper modules.

    +--------------+     +-------------+
    |              |     |             |
    | path_helpers |     | url_helpers | (named routes added here)
    |              |     |             |
    +------+----+--+     +----+--+-----+
           ^    ^             ^  ^
           |    |             |  |
           |  +---------------+  |
           |  | |                |
           |  | +-----------+    |
           |  |             |    |
    +------+--+---+      +--+----+-----+
    |             |      |             |
    | (singleton) +------+ url_helpers | (duplicated for each controller)
    |             |      |             |
    +-------------+      +-----+-------+
                               ^
                               |
                               |
                     +---------+-------+
                     |                 |
                     | UsersController |
                     |                 |
                     +-----------------+

The result of this is that when named routes were added to the two
top-level modules, defining those methods became extremely slow because
Ruby has to invalidate the class method caches on all of the generated
modules as well as each controller. Furthermore because there were
multiple paths up the inheritance chain (ex. one through
UsersController's generated module and one through
ApplicationController's genereated module) it would end up invaliding
the classes multiple times (ideally Ruby would be smarter about this,
but it's much easier to solve in Rails).

Caching this module means that all controllers should include the same
module and defining these named routes should be 3-4x faster.

This method can generate two different modules, based on whether
supports_path is truthy, so those are cached separately.

rafaelfranca commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -479,7 +480,7 @@ def #{name}
         routes = self
 
-        Module.new do
+        @url_helpers_cache[!!supports_path] ||= Module.new do
```
Should we use a hash or just define two separated instance variables like it was when we used to have this cache?

jhawthorn replied:
👍 I switched to the previous structure and method naming.

rafaelfranca reviewed (commented):
This cache was removed in https://github.com/rails/rails/commit/0acd4a57768fc3c7e758f9f4b26563797f43e7ef. @eileencodes do you remember why?

Closes #24554 as well.

jhawthorn committed:
Add regression test for url_helpers duplication

This adds a regression test that previously failed checking controllers
don't have multiple ancestors which include url_helpers or path_helpers.

This is checked by digging through the ancestors and seeing if they
include either the url_helpers_module or path_helpers_module and making
sure only one ancestor does that.

We usually don't test performance changes (and also would not prefer to
test implementation this closely), but because this has regressed in the
past, and I think it would be relatively easy to accidentally introduce
again, I think we should test this one.

casperisfine commented:
> I'd love to know if this helps you too

I just ran the profiles, this has a massive effect:

Before:
```
boot time 1m24
==================================
  Mode: wall(1000)
  Samples: 58959 (22.91% miss rate)
  GC: 18466 (31.32%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     18466  (31.3%)       18466  (31.3%)     (garbage collection)
      7316  (12.4%)        7316  (12.4%)     ActionDispatch::Routing::RouteSet::NamedRouteCollection#define_url_helper
```

After:
```
boot time: 56s
==================================
  Mode: wall(1000)
  Samples: 52215 (7.13% miss rate)
  GC: 16824 (32.22%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     16824  (32.2%)       16824  (32.2%)     (garbage collection)
      5036   (9.6%)        5036   (9.6%)     ActionDispatch::Routing::RouteSet::NamedRouteCollection#define_url_helper
```

So 💯 for this change.

One small downside but that IMO is a good tradeoff, is a 2.12MB increase in resident memory.

I wonder if that cache could be dropped after it's no longer useful if  eager_loading / cache_classes is enabled.

eileencodes commented:
I think it was just we found another way to get more or the same savings without the cache. Our benchmarks were against integration test so IIRC at the time we didn't see a win from the cache for those measurements.

jhawthorn commented:
> One small downside but that IMO is a good tradeoff, is a 2.12MB increase in resident memory.

That's odd. I'd expect this to use less memory if anything (but not significantly). Is it possible that's just an inconsistency? We should now only be generating two of these modules and neither of them should have previously had a chance to be GC'd (one is included on controllers, the other on mailers).

I'd also love to know if https://github.com/ruby/ruby/pull/2752 improves performance any more 😁

casperisfine commented:
> Is it possible that's just an inconsistency?

Yes, it's very well possible. Our app have plenty of caches, so it can show bigger memory usage because it re-used some semi-stale cache.

Looking at the diff, I think it's fair to assume it's an inconsistency.

inopinatus commented:
I've noticed a regression involving application subclasses of isolated engine controllers seeing the wrong route helpers when they share a common ancestor (most likely ApplicationController).

I think that's a common pattern in the wild, may even be the default for some very popular authentication engines. I'll open a PR with a failing test case to illustrate.
